drivers/node/aws/aws.go

```diff
@@ -0,0 +1,128 @@
+	// AwsRegion is the region of aws ec2
+	AwsRegion = "us-east-1"
+)
+
+type aws struct {
+	node.Driver
+	session     *session.Session
+	credentials *credentials.Credentials
+	config      *aws_pkg.Config
+	svc         *ec2.EC2
+	Instances   []*ec2.Instance
+}
+
+func (aws *aws) String() string {
+	return DriverName
+}
+
+func (aws *aws) Init(sched string) error {
+	sess := session.Must(session.NewSession())
+	aws.session = sess
+	creds := credentials.NewEnvCredentials()
+	aws.credentials = creds
+	config := &aws_pkg.Config{Region: aws_pkg.String(AwsRegion)}
+	config.WithCredentials(creds)
+	aws.config = config
+	svc := ec2.New(sess, config)
+	aws.svc = svc
+	//Get all nodes from east region
+	instances := aws.getAllInstancesInEast()
+	aws.Instances = instances
+	return nil
+}
+
+func (aws *aws) TestConnection(n node.Node, options node.TestConectionOpts) error {
+	return nil
+}
+
+func (aws *aws) RebootNode(n node.Node, options node.RebootNodeOpts) error {
+	var err error
+	_, err = aws.getAddrToConnect(n)
+	if err != nil {
+		return &ErrFailedToRebootNode{
+			Node:  n,
+			Cause: fmt.Sprintf("failed to get node address due to: %v", err),
+		}
+	}
+	instanceID, err := aws.getNodeIDByPrivAddr(n)
+	if err != nil {
+		fmt.Println("there was an error get instance ID by its private IP ", n.Addresses[0], err.Error())
+		log.Fatal(err.Error())
+	}
+	//Reboot the instance by its InstanceID
+	rebootInput := &ec2.RebootInstancesInput{
+		InstanceIds: []*string{
+			aws_pkg.String(instanceID),
+		},
+	}
+	fmt.Printf("Now reboot instance %+v\n", rebootInput)
+	_, err = aws.svc.RebootInstances(rebootInput)
+	if err != nil {
+		return fmt.Errorf("there was an error reboot instances %s, %q", instanceID, err.Error())
+	}
+	return nil
+}
+
+func (aws *aws) ShutdownNode(n node.Node, options node.ShutdownNodeOpts) error {
+	return nil
+}
+
+func (aws *aws) getAddrToConnect(n node.Node) (string, error) {
+	if n.Addresses == nil || len(n.Addresses) == 0 {
+		return "", fmt.Errorf("no address available to connect")
+	}
+	addr := n.Addresses[0] // TODO don't stick to first address
+	return addr, nil
+}
+
+func (aws *aws) getAllInstancesInEast() []*ec2.Instance {
+	params := &ec2.DescribeInstancesInput{}
+	resp, err := aws.svc.DescribeInstances(params)
+	if err != nil {
+		fmt.Println("there was an error listing instances in", AwsRegion, err.Error())
+		log.Fatal(err.Error())
+	}
+	reservations := resp.Reservations
+	//If resp is empty, then return empty []
+	instances := []*ec2.Instance{}
+	for _, resv := range reservations {
+		for _, ins := range resv.Instances {
+			instances = append(instances, ins)
+		}
+	}
+	fmt.Printf("Found %d instances in %s\n", len(instances), AwsRegion)
+	return instances
+}
+
+func (aws *aws) getNodeIDByPrivAddr(n node.Node) (string, error) {
+	fmt.Printf("Get instance ID of node whose private ip is %s\n", n.Addresses[0])
+	for _, i := range aws.Instances {
+		if aws_pkg.StringValue(i.PrivateIpAddress) == n.Addresses[0] {
```

Comment:
Fragile code. We cannot simply compare to `n.Addresses[0]`. Since the AWS private IP can easily be the 2nd, 3rd, ... address. Compare with all `n.Addresses`.